File: im-desktop/src/services/identityFingerprintService.ts
/**
 * 身份指纹验证服务
 * 处理安全码验证、二维码扫描、密钥变更通知等客户端逻辑
 */

import { apiClient } from './apiClient';
import { websocketService } from './websocketService';

export interface FingerprintVerification {
  id: number;
  type: 'SAFETY_CODE' | 'QR_SCAN' | 'KEY_CHANGE' | 'DEVICE_FINGERPRINT' | 'BIOMETRIC' | 'BACKUP_CODE';
  status: 'PENDING' | 'VERIFIED' | 'EXPIRED' | 'FAILED' | 'REVOKED' | 'LOCKED';
  verificationCode?: string;
  expiresAt?: string;
  verifiedAt?: string;
  deviceId?: string;
  deviceName?: string;
  deviceType?: 'DESKTOP' | 'MOBILE' | 'WEB' | 'ANY';
  qrImageUrl?: string;
  createdAt: string;
  updatedAt: string;
}

export interface SafetyCodeRequest {
  deviceId?: string;
  deviceName?: string;
  ipAddress?: string;
  userAgent?: string;
}

export interface SafetyCodeVerify {
  verificationCode: string;
  deviceId: string;
}

export interface QrCodeRequest {
  deviceId: string;
  deviceName: string;
}

export interface QrCodeVerify {
  qrData: string;
  scanningDeviceId: string;
}

export interface KeyChangeNotify {
  deviceId: string;
  keyType: string;
  changeReason: string;
}

export interface VerificationStats {
  userId: number;
  totalVerifications: number;
  successfulVerifications: number;
  failedVerifications: number;
  lastVerification: string;
  preferredMethod: string;
  devicesUsed: number;
}

export interface VerificationMethod {
  method: string;
  name: string;
  description: string;
  supportedPlatforms: string[];
  expirationMinutes?: number;
  requiresNetwork?: boolean;
  requiresConfirmation?: boolean;
  automatic?: boolean;
}

export class IdentityFingerprintService {
  private static instance: IdentityFingerprintService;
  
  public static getInstance(): IdentityFingerprintService {
    if (!IdentityFingerprintService.instance) {
      IdentityFingerprintService.instance = new IdentityFingerprintService();
    }
    return IdentityFingerprintService.instance;
  }
  
  /**
   * 请求安全码验证
   */
  async requestSafetyCodeVerification(params: SafetyCodeRequest): Promise<{
    success: boolean;
    fingerprintId: number;
    verificationCode: string;
    expiresAt: string;
  }> {
    try {
      const response = await apiClient.post('/security/fingerprint/safety-code/request', params);
      return response.data;
    } catch (error) {
      console.error('请求安全码验证失败:', error);
      throw new Error('请求安全码验证失败，请检查网络连接');
    }
  }
  
  /**
   * 验证安全码
   */
  async verifySafetyCode(params: SafetyCodeVerify): Promise<{
    success: boolean;
    fingerprintId: number;
    verifiedAt: string;
    status: string;
  }> {
    try {
      const response = await apiClient.post('/security/fingerprint/safety-code/verify', params);
      return response.data;
    } catch (error: any) {
      console.error('验证安全码失败:', error);
      if (error.response?.data?.error) {
        throw new Error(error.response.data.error);
      }
      throw new Error('验证安全码失败，请重试');
    }
  }
  
  /**
   * 请求二维码验证
   */
  async requestQrCodeVerification(params: QrCodeRequest): Promise<{
    success: boolean;
    fingerprintId: number;
    qrImageUrl: string;
    expiresAt: string;
  }> {
    try {
      const response = await apiClient.post('/security/fingerprint/qr-code/request', params);
      return response.data;
    } catch (error) {
      console.error('请求二维码验证失败:', error);
      throw new Error('请求二维码验证失败');
    }
  }
  
  /**
   * 验证二维码扫描
   */
  async verifyQrCodeScan(params: QrCodeVerify): Promise<{
    success: boolean;
    fingerprintId: number;
    verifiedAt: string;
    status: string;
  }> {
    try {
      const response = await apiClient.post('/security/fingerprint/qr-code/verify', params);
      return response.data;
    } catch (error: any) {
      console.error('验证二维码扫描失败:', error);
      if (error.response?.data?.error) {
        throw new Error(error.response.data.error);
      }
      throw new Error('验证二维码扫描失败');
    }
  }
  
  /**
   * 发送密钥变更通知
   */
  async notifyKeyChange(params: KeyChangeNotify): Promise<{
    success: boolean;
    fingerprintId: number;
    createdAt: string;
    requiresConfirmation: boolean;
  }> {
    try {
      const response = await apiClient.post('/security/fingerprint/key-change/notify', params);
      return response.data;
    } catch (error) {
      console.error('发送密钥变更通知失败:', error);
      throw new Error('发送密钥变更通知失败');
    }
  }
  
  /**
   * 获取验证历史
   */
  async getVerificationHistory(limit: number = 10): Promise<{
    success: boolean;
    userId: number;
    total: number;
    history: Array<{
      id: number;
      type: string;
      status: string;
      verifiedAt: string;
      deviceName: string;
      deviceType: string;
    }>;
  }> {
    try {
      const response = await apiClient.get('/security/fingerprint/history', {
        params: { limit }
      });
      return response.data;
    } catch (error) {
      console.error('获取验证历史失败:', error);
      throw new Error('获取验证历史失败');
    }
  }
  
  /**
   * 获取待处理验证
   */
  async getPendingVerifications(): Promise<{
    success: boolean;
    userId: number;
    count: number;
    pendingVerifications: Array<{
      id: number;
      type: string;
      createdAt: string;
      expiresAt: string;
      deviceName: string;
      qrImageUrl?: string;
      hasVerificationCode?: boolean;
    }>;
  }> {
    try {
      const response = await apiClient.get('/security/fingerprint/pending');
      return response.data;
    } catch (error) {
      console.error('获取待处理验证失败:', error);
      throw new Error('获取待处理验证失败');
    }
  }
  
  /**
   * 撤销指纹验证
   */
  async revokeFingerprint(fingerprintId: number, reason: string): Promise<{
    success: boolean;
    fingerprintId: number;
    revokedAt: string;
    reason: string;
  }> {
    try {
      const response = await apiClient.post(`/security/fingerprint/${fingerprintId}/revoke`, null, {
        params: { reason }
      });
      return response.data;
    } catch (error) {
      console.error('撤销指纹验证失败:', error);
      throw new Error('撤销指纹验证失败');
    }
  }
  
  /**
   * 获取验证统计
   */
  async getVerificationStats(): Promise<{
    success: boolean;
    stats: VerificationStats;
  }> {
    try {
      const response = await apiClient.get('/security/fingerprint/stats');
      return response.data;
    } catch (error) {
      console.error('获取验证统计失败:', error);
      throw new Error('获取验证统计失败');
    }
  }
  
  /**
   * 检查验证状态
   */
  async checkVerificationStatus(fingerprintId: number): Promise<{
    success: boolean;
    status: {
      fingerprintId: number;
      status: string;
      verifiedAt: string;
      method: string;
      device: string;
    };
  }> {
    try {
      const response = await apiClient.get(`/security/fingerprint/${fingerprintId}/status`);
      return response.data;
    } catch (error) {
      console.error('检查验证状态失败:', error);
      throw new Error('检查验证状态失败');
    }
  }
  
  /**
   * 重新发送验证码
   */
  async resendVerificationCode(fingerprintId: number): Promise<{
    success: boolean;
    fingerprintId: number;
    resentAt: string;
    message: string;
  }> {
    try {
      const response = await apiClient.post(`/security/fingerprint/${fingerprintId}/resend`);
      return response.data;
    } catch (error) {
      console.error('重新发送验证码失败:', error);
      throw new Error('重新发送验证码失败');
    }
  }
  
  /**
   * 获取支持的验证方法
   */
  async getSupportedMethods(): Promise<{
    success: boolean;
    methods: VerificationMethod[];
  }> {
    try {
      const response = await apiClient.get('/security/fingerprint/methods');
      return response.data;
    } catch (error) {
      console.error('获取支持的验证方法失败:', error);
      throw new Error('获取支持的验证方法失败');
    }
  }
  
  /**
   * 健康检查
   */
  async healthCheck(): Promise<{
    service: string;
    status: string;
    timestamp: string;
    version: string;
  }> {
    try {
      const response = await apiClient.get('/security/fingerprint/health');
      return response.data;
    } catch (error) {
      console.error('身份指纹验证服务健康检查失败:', error);
      throw new Error('身份指纹验证服务不可用');
    }
  }
  
  /**
   * 监听WebSocket安全事件
   */
  setupWebSocketListeners() {
    // 监听指纹验证成功事件
    websocketService.subscribe('security', (data: any) => {
      if (data.type === 'FINGERPRINT_VERIFIED') {
        this.handleFingerprintVerified(data);
      } else if (data.type === 'QR_CODE_SCANNED') {
        this.handleQrCodeScanned(data);
      }
    });
  }
  
  /**
   * 处理指纹验证成功事件
   */
  private handleFingerprintVerified(data: any) {
    console.log('指纹验证成功:', data);
    
    // 触发自定义事件
    const event = new CustomEvent('fingerprint-verified', {
      detail: data
    });
    window.dispatchEvent(event);
    
    // 可以在这里更新UI状态
    // 例如：显示验证成功的通知
    this.showNotification('身份验证成功', '您的身份验证已成功完成', 'success');
  }
  
  /**
   * 处理二维码扫描事件
   */
  private handleQrCodeScanned(data: any) {
    console.log('二维码被扫描:', data);
    
    const event = new CustomEvent('qr-code-scanned', {
      detail: data
    });
    window.dispatchEvent(event);
    
    this.showNotification('设备扫描成功', `设备 ${data.scanningDeviceId} 扫描了您的二维码`, 'info');
  }
  
  /**
   * 显示通知
   */
  private showNotification(title: string, message: string, type: 'success' | 'error' | 'info' | 'warning') {
    // 这里可以集成通知系统
    console.log(`[${type.toUpperCase()}] ${title}: ${message}`);
    
    // 简单实现：使用浏览器的通知API
    if ('Notification' in window && Notification.permission === 'granted') {
      new Notification(title, {
        body: message,
        icon: type === 'success' ? '/icons/success.png' : 
               type === 'error' ? '/icons/error.png' :
               type === 'warning' ? '/icons/warning.png' : '/icons/info.png'
      });
    }
  }
  
  /**
   * 生成设备ID
   */
  generateDeviceId(): string {
    // 生成基于浏览器指纹的设备ID
    const userAgent = navigator.userAgent;
    const language = navigator.language;
    const timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    const screenResolution = `${window.screen.width}x${window.screen.height}`;
    
    // 简单的哈希算法
    const data = `${userAgent}-${language}-${timezone}-${screenResolution}`;
    let hash = 0;
    
    for (let i = 0; i < data.length; i++) {
      const char = data.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // 转换为32位整数
    }
    
    return `device_${Math.abs(hash)}`;
  }
  
  /**
   * 获取设备信息
   */
  getDeviceInfo(): {
    deviceId: string;
    deviceName: string;
    deviceType: string;
    userAgent: string;
    ipAddress?: string;
  } {
    const deviceId = this.generateDeviceId();
    const deviceName = this.getDeviceName();
    const deviceType = this.getDeviceType();
    const userAgent = navigator.userAgent;
    
    return {
      deviceId,
      deviceName,
      deviceType,
      userAgent
    };
  }
  
  /**
   * 获取设备名称
   */
  private getDeviceName(): string {
    // 尝试从本地存储获取设备名称
    const savedName = localStorage.getItem('device_name');
    if (savedName) {
      return savedName;
    }
    
    // 根据用户代理推断设备名称
    const ua = navigator.userAgent;
    
    if (ua.includes('Windows')) {
      return 'Windows Desktop';
    } else if (ua.includes('Mac')) {
      return 'Mac Desktop';
    } else if (ua.includes('Linux')) {
      return 'Linux Desktop';
    } else if (ua.includes('Android')) {
      return 'Android Device';
    } else if (ua.includes('iPhone') || ua.includes('iPad')) {
      return 'iOS Device';
    } else {
      return 'Web Browser';
    }
  }
  
  /**
   * 获取设备类型
   */
  private getDeviceType(): string {
    const ua = navigator.userAgent;
    
    if (ua.includes('Mobile') || ua.includes('Android') || ua.includes('iPhone') || ua.includes('iPad')) {
      return 'MOBILE';
    } else {
      return 'DESKTOP';
    }
  }
  
  /**
   * 保存设备名称
   */
  saveDeviceName(name: string): void {
    localStorage.setItem('device_name', name);
  }
  
  /**
   * 生成二维码数据
   */
  generateQrData(): string {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2, 10);
    return `IM_QR_${timestamp}_${random}`;
  }
  
  /**
   * 验证二维码数据格式
   */
  validateQrData(qrData: string): boolean {
    // 检查二维码数据格式
    return qrData.startsWith('IM_QR_') || qrData.startsWith('IM_VERIFY_');
  }
}

export const identityFingerprintService = IdentityFingerprintService.getInstance();

File: im-mobile/lib/services/identity_fingerprint_service.dart
/// 身份指纹验证服务
/// 处理安全码验证、二维码扫描、密钥变更通知等移动端逻辑

import 'dart:convert';
import 'dart:math';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:device_info_plus/device_info_plus.dart';
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

/// 指纹验证实体
class FingerprintVerification {
  final int id;
  final String type;
  final String status;
  final String? verificationCode;
  final DateTime? expiresAt;
  final DateTime? verifiedAt;
  final String? deviceId;
  final String? deviceName;
  final String? deviceType;
  final String? qrImageUrl;
  final DateTime createdAt;
  final DateTime updatedAt;

  FingerprintVerification({
    required this.id,
    required this.type,
    required this.status,
    this.verificationCode,
    this.expiresAt,
    this.verifiedAt,
    this.deviceId,
    this.deviceName,
    this.deviceType,
    this.qrImageUrl,
    required this.createdAt,
    required this.updatedAt,
  });

  factory FingerprintVerification.fromJson(Map<String, dynamic> json) {
    return FingerprintVerification(
      id: json['id'] as int,
      type: json['type'] as String,
      status: json['status'] as String,
      verificationCode: json['verificationCode'],
      expiresAt: json['expiresAt'] != null ? DateTime.parse(json['expiresAt']) : null,
      verifiedAt: json['verifiedAt'] != null ? DateTime.parse(json['verifiedAt']) : null,
      deviceId: json['deviceId'],
      deviceName: json['deviceName'],
      deviceType: json['deviceType'],
      qrImageUrl: json['qrImageUrl'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type,
      'status': status,
      'verificationCode': verificationCode,
      'expiresAt': expiresAt?.toIso8601String(),
      'verifiedAt': verifiedAt?.toIso8601String(),
      'deviceId': deviceId,
      'deviceName': deviceName,
      'deviceType': deviceType,
      'qrImageUrl': qrImageUrl,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

/// 安全码请求参数
class SafetyCodeRequest {
  final String? deviceId;
  final String? deviceName;
  final String? ipAddress;
  final String? userAgent;

  SafetyCodeRequest({
    this.deviceId,
    this.deviceName,
    this.ipAddress,
    this.userAgent,
  });

  Map<String, dynamic> toJson() {
    return {
      if (deviceId != null) 'deviceId': deviceId,
      if (deviceName != null) 'deviceName': deviceName,
      if (ipAddress != null) 'ipAddress': ipAddress,
      if (userAgent != null) 'userAgent': userAgent,
    };
  }
}

/// 安全码验证参数
class SafetyCodeVerify {
  final String verificationCode;
  final String deviceId;

  SafetyCodeVerify({
    required this.verificationCode,
    required this.deviceId,
  });

  Map<String, dynamic> toJson() {
    return {
      'verificationCode': verificationCode,
      'deviceId': deviceId,
    };
  }
}

/// 二维码请求参数
class QrCodeRequest {
  final String deviceId;
  final String deviceName;

  QrCodeRequest({
    required this.deviceId,
    required this.deviceName,
  });

  Map<String, dynamic> toJson() {
    return {
      'deviceId': deviceId,
      'deviceName': deviceName,
    };
  }
}

/// 二维码验证参数
class QrCodeVerify {
  final String qrData;
  final String scanningDeviceId;

  QrCodeVerify({
    required this.qrData,
    required this.scanningDeviceId,
  });

  Map<String, dynamic> toJson() {
    return {
      'qrData': qrData,
      'scanningDeviceId': scanningDeviceId,
    };
  }
}

/// 密钥变更通知参数
class KeyChangeNotify {
  final String deviceId;
  final String keyType;
  final String changeReason;

  KeyChangeNotify({
    required this.deviceId,
    required this.keyType,
    required this.changeReason,
  });

  Map<String, dynamic> toJson() {
    return {
      'deviceId': deviceId,
      'keyType': keyType,
      'changeReason': changeReason,
    };
  }
}

/// 身份指纹验证服务
class IdentityFingerprintService {
  static final IdentityFingerprintService _instance = IdentityFingerprintService._internal();
  
  factory IdentityFingerprintService() {
    return _instance;
  }
  
  IdentityFingerprintService._internal();
  
  final String _baseUrl = 'https://api.example.com/api/security/fingerprint';
  late String _authToken;
  final DeviceInfoPlugin _deviceInfo = DeviceInfoPlugin();
  
  /// 初始化服务
  Future<void> initialize(String authToken) async {
    _authToken = authToken;
    await _setupDeviceInfo();
  }
  
  /// 请求安全码验证
  Future<Map<String, dynamic>> requestSafetyCodeVerification(SafetyCodeRequest request) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/safety-code/request'),
        headers: _buildHeaders(),
        body: jsonEncode(request.toJson()),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('请求安全码验证失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('请求安全码验证失败: $e');
      rethrow;
    }
  }
  
  /// 验证安全码
  Future<Map<String, dynamic>> verifySafetyCode(SafetyCodeVerify verify) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/safety-code/verify'),
        headers: _buildHeaders(),
        body: jsonEncode(verify.toJson()),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        final errorData = jsonDecode(response.body);
        throw Exception(errorData['error'] ?? '验证安全码失败');
      }
    } catch (e) {
      debugPrint('验证安全码失败: $e');
      rethrow;
    }
  }
  
  /// 请求二维码验证
  Future<Map<String, dynamic>> requestQrCodeVerification(QrCodeRequest request) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/qr-code/request'),
        headers: _buildHeaders(),
        body: jsonEncode(request.toJson()),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('请求二维码验证失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('请求二维码验证失败: $e');
      rethrow;
    }
  }
  
  /// 验证二维码扫描
  Future<Map<String, dynamic>> verifyQrCodeScan(QrCodeVerify verify) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/qr-code/verify'),
        headers: _buildHeaders(),
        body: jsonEncode(verify.toJson()),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        final errorData = jsonDecode(response.body);
        throw Exception(errorData['error'] ?? '验证二维码扫描失败');
      }
    } catch (e) {
      debugPrint('验证二维码扫描失败: $e');
      rethrow;
    }
  }
  
  /// 发送密钥变更通知
  Future<Map<String, dynamic>> notifyKeyChange(KeyChangeNotify notify) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/key-change/notify'),
        headers: _buildHeaders(),
        body: jsonEncode(notify.toJson()),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('发送密钥变更通知失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('发送密钥变更通知失败: $e');
      rethrow;
    }
  }
  
  /// 获取验证历史
  Future<Map<String, dynamic>> getVerificationHistory({int limit = 10}) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/history').replace(queryParameters: {'limit': limit.toString()}),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('获取验证历史失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('获取验证历史失败: $e');
      rethrow;
    }
  }
  
  /// 获取待处理验证
  Future<Map<String, dynamic>> getPendingVerifications() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/pending'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('获取待处理验证失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('获取待处理验证失败: $e');
      rethrow;
    }
  }
  
  /// 撤销指纹验证
  Future<Map<String, dynamic>> revokeFingerprint(int fingerprintId, String reason) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/$fingerprintId/revoke').replace(queryParameters: {'reason': reason}),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('撤销指纹验证失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('撤销指纹验证失败: $e');
      rethrow;
    }
  }
  
  /// 获取验证统计
  Future<Map<String, dynamic>> getVerificationStats() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/stats'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('获取验证统计失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('获取验证统计失败: $e');
      rethrow;
    }
  }
  
  /// 检查验证状态
  Future<Map<String, dynamic>> checkVerificationStatus(int fingerprintId) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/$fingerprintId/status'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('检查验证状态失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('检查验证状态失败: $e');
      rethrow;
    }
  }
  
  /// 重新发送验证码
  Future<Map<String, dynamic>> resendVerificationCode(int fingerprintId) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/$fingerprintId/resend'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('重新发送验证码失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('重新发送验证码失败: $e');
      rethrow;
    }
  }
  
  /// 获取支持的验证方法
  Future<Map<String, dynamic>> getSupportedMethods() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/methods'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('获取支持的验证方法失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('获取支持的验证方法失败: $e');
      rethrow;
    }
  }
  
  /// 健康检查
  Future<Map<String, dynamic>> healthCheck() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/health'),
        headers: _buildHeaders(),
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('健康检查失败: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('健康检查失败: $e');
      rethrow;
    }
  }
  
  /// 生成设备ID
  Future<String> generateDeviceId() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final savedDeviceId = prefs.getString('device_id');
      
      if (savedDeviceId != null) {
        return savedDeviceId;
      }
      
      // 生成基于设备信息的设备ID
      String deviceInfoString;
      
      if (defaultTargetPlatform == TargetPlatform.android) {
        final androidInfo = await _deviceInfo.androidInfo;
        deviceInfoString = '${androidInfo.model}-${androidInfo.id}-${androidInfo.brand}';
      } else if (defaultTargetPlatform == TargetPlatform.iOS) {
        final iosInfo = await _deviceInfo.iosInfo;
        deviceInfoString = '${iosInfo.model}-${iosInfo.identifierForVendor}';
      } else {
        deviceInfoString = '${DateTime.now().millisecondsSinceEpoch}-${Random().nextInt(1000000)}';
      }
      
      // 简单哈希
      var hash = 0;
      for (var i = 0; i < deviceInfoString.length; i++) {
        final char = deviceInfoString.codeUnitAt(i);
        hash = ((hash << 5) - hash) + char;
        hash = hash & hash; // 转换为32位整数
      }
      
      final deviceId = 'mobile_device_${hash.abs()}';
      await prefs.setString('device_id', deviceId);
      
      return deviceId;
    } catch (e) {
      debugPrint('生成设备ID失败: $e');
      // 使用备用方法
      return 'mobile_device_${DateTime.now().millisecondsSinceEpoch}_${Random().nextInt(1000000)}';
    }
  }
  
  /// 获取设备信息
  Future<Map<String, dynamic>> getDeviceInfo() async {
    try {
      final deviceId = await generateDeviceId();
      final prefs = await SharedPreferences.getInstance();
      final deviceName = prefs.getString('device_name') ?? await _getDefaultDeviceName();
      final deviceType = 'MOBILE';
      
      // 获取网络信息
      final connectivity = await Connectivity().checkConnectivity();
      final networkType = connectivity.name;
      
      return {
        'deviceId': deviceId,
        'deviceName': deviceName,
        'deviceType': deviceType,
        'networkType': networkType,
        'platform': defaultTargetPlatform.name,
      };
    } catch (e) {
      debugPrint('获取设备信息失败: $e');
      return {
        'deviceId': 'unknown_device',
        'deviceName': 'Mobile Device',
        'deviceType': 'MOBILE',
        'networkType': 'unknown',
        'platform': 'unknown',
      };
    }
  }
  
  /// 保存设备名称
  Future<void> saveDeviceName(String name) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString('device_name', name);
    } catch (e) {
      debugPrint('保存设备名称失败: $e');
    }
  }
  
  /// 生成验证码
  String generateVerificationCode() {
    final random = Random();
    return '${random.nextInt(9)+1}${random.nextInt(10)}${random.nextInt(10)}${random.nextInt(10)}${random.nextInt(10)}${random.nextInt(10)}';
  }
  
  /// 生成二维码数据
  String generateQrData() {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final random = Random().nextInt(100000000);
    return 'IM_QR_${timestamp}_$random';
  }
  
  /// 验证二维码数据格式
  bool validateQrData(String qrData) {
    return qrData.startsWith('IM_QR_') || qrData.startsWith('IM_VERIFY_');
  }
  
  /// 构建请求头
  Map<String, String> _buildHeaders() {
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $_authToken',
      'User-Agent': 'IM-Mobile-App',
    };
  }
  
  /// 设置设备信息
  Future<void> _setupDeviceInfo() async {
    try {
      // 确保设备ID已生成
      await generateDeviceId();
    } catch (e) {
      debugPrint('设置设备信息失败: $e');
    }
  }
  
  /// 获取默认设备名称
  Future<String> _getDefaultDeviceName() async {
    try {
      if (defaultTargetPlatform == TargetPlatform.android) {
        final androidInfo = await _deviceInfo.androidInfo;
        return '${androidInfo.brand} ${androidInfo.model}';
      } else if (defaultTargetPlatform == TargetPlatform.iOS) {
        final iosInfo = await _deviceInfo.iosInfo;
        return '${iosInfo.model}';
      } else {
        return 'Mobile Device';
      }
    } catch (e) {
      debugPrint('获取默认设备名称失败: $e');
      return 'Mobile Device';
    }
  }
  
  /// 显示通知（简化版本）
  void showNotification(String title, String message, String type) {
    debugPrint('[$type] $title: $message');
    
    // 在实际应用中，这里应该集成Flutter的通知插件
    // 例如：flutter_local_notifications
  }
}

// 全局服务实例
final identityFingerprintService = IdentityFingerprintService();